godot-core/src/obj/raw.rs

```diff
@@ -0,0 +1,624 @@
+        // virtual inheritance etc.). It *seems* to work so far, but this is no indication it's not UB.
+        //
+        // The Deref/DerefMut impls for T implement an "implicit upcast" on the object (not Gd) level and
+        // rely on this (e.g. &Node3D -> &Node).
+
+        let result = unsafe { self.ffi_cast::<U>() };
+        match result {
+            Some(cast_obj) => {
+                // duplicated ref, one must be wiped
+                std::mem::forget(self);
+                Ok(cast_obj)
+            }
+            None => Err(self),
+        }
+    }
+
+    /// # Safety
+    /// Does not transfer ownership and is thus unsafe. Also operates on shared ref. Either the parameter or
+    /// the return value *must* be forgotten (since reference counts are not updated).
+    pub(super) unsafe fn ffi_cast<U>(&self) -> Option<RawGd<U>>
+    where
+        U: GodotClass,
+    {
+        if self.is_null() {
+            // Null can be cast to anything.
+            // Forgetting a null doesn't do anything, since dropping a null also does nothing.
+            return Some(RawGd::new_null());
+        }
+
+        let class_tag = interface_fn!(classdb_get_class_tag)(U::class_name().string_sys());
+        let cast_object_ptr = interface_fn!(object_cast_to)(self.obj_sys(), class_tag);
+
+        // Create weak object, as ownership will be moved and reference-counter stays the same
+        sys::ptr_then(cast_object_ptr, |ptr| RawGd::from_obj_sys_weak(ptr))
+    }
+
+    pub(crate) fn as_ref_counted<R>(&self, apply: impl Fn(&mut engine::RefCounted) -> R) -> R {
+        debug_assert!(
+            self.is_instance_valid(),
+            "as_ref_counted() on freed instance; maybe forgot to increment reference count?"
+        );
+
+        let tmp = unsafe { self.ffi_cast::<engine::RefCounted>() };
+        let mut tmp = tmp.expect("object expected to inherit RefCounted");
+        let return_val =
+            <engine::RefCounted as GodotClass>::Declarer::scoped_mut(&mut tmp, |obj| apply(obj));
+
+        std::mem::forget(tmp); // no ownership transfer
+        return_val
+    }
+
+    pub(crate) fn as_object<R>(&self, apply: impl Fn(&mut engine::Object) -> R) -> R {
+        // Note: no validity check; this could be called by to_string(), which can be called on dead instances
+
+        let tmp = unsafe { self.ffi_cast::<engine::Object>() };
+        let mut tmp = tmp.expect("object expected to inherit Object; should never fail");
+        // let return_val = apply(tmp.inner_mut());
+        let return_val =
+            <engine::Object as GodotClass>::Declarer::scoped_mut(&mut tmp, |obj| apply(obj));
+
+        std::mem::forget(tmp); // no ownership transfer
+        return_val
+    }
+
+    // Target is always an engine class:
+    // * if T is an engine class => T
+    // * if T is a user class => T::Base
+    pub(super) fn as_target(
+        &self,
+    ) -> Option<&<<T as GodotClass>::Declarer as dom::Domain>::DerefTarget<T>> {
```

Comment:
Is it possible to create a type alias (outside), e.g.
```rs
type GdDerefTarget<T> = <<T as GodotClass>::Declarer as dom::Domain>::DerefTarget<T>;
```

This occurs four times starting here, also in `as_target_mut`.

Before, this was the `Deref` trait, so `Self::Target` took that role.